src/libsyntax/parse/parser.rs

```diff
@@ -3210,7 +3228,23 @@ impl<'a> Parser<'a> {
             err.span_label(sp, "expected if condition here");
             return Err(err)
         }
-        let thn = self.parse_block()?;
+        let not_block = self.token != token::OpenDelim(token::Brace);
+        let fat_arrow_sp = if self.token == token::FatArrow {
+            Some(self.span)
+        } else {
+            None
+        };
+        let thn = self.parse_block().map_err(|mut err| {
+            if let Some(sp) = fat_arrow_sp {
+                // if cond => expr
```

Comment:
Is there an issue about this or any other confirmation that this mistake is realistic?

Reply:
There was an [existing test case](https://github.com/estebank/rust/commit/72d629caa50fa482dd4165cb43b9fb02ac24c8d8#diff-6ca7ee6ac85a875788a20c56dda8a64d), but it is indeed a big corner case. I can remove the special diagnostic.

Reply:
Yes, I think just removing this would be for the better.